modules/apps/27-interchain-accounts/keeper/handshake.go

```diff
@@ -84,23 +87,37 @@ func (k Keeper) OnChanOpenTry(
 		return err
 	}
 
+	accAddr := types.GenerateAddress(counterparty.PortId)
+	parsedAddr := types.ParseAddressFromVersion(version)
+	if parsedAddr != accAddr.String() {
+		return sdkerrors.Wrapf(types.ErrInvalidAccountAddress, "version contains invalid account address: expected %s, got %s", parsedAddr, accAddr)
+	}
+
 	// Register interchain account if it does not already exist
-	k.RegisterInterchainAccount(ctx, counterparty.PortId)
+	k.RegisterInterchainAccount(ctx, accAddr, counterparty.PortId)
+
 	return nil
 }
 
+// OnChanOpenAck sets the active channel for the interchain account/owner pair
+// and stores the associated interchain account address in state keyed by it's corresponding port identifier
+//
+// Controller Chain
 func (k Keeper) OnChanOpenAck(
 	ctx sdk.Context,
 	portID,
 	channelID string,
 	counterpartyVersion string,
 ) error {
-	if counterpartyVersion != types.Version {
-		return sdkerrors.Wrapf(types.ErrInvalidVersion, "invalid counterparty version: %s, expected %s", counterpartyVersion, types.Version)
+	if err := types.ValidateVersion(counterpartyVersion); err != nil {
```

Comment:
@colin-axner I'm just wondering is there any way that a malicious relayer could pass in an address here to the ack step that isn't the address we validated on the `Try` step? I think we already discussed this but just refreshing my brain after the weekend.

Reply:
No, core IBC verifies a proof of the channel state set on the [counterparty](https://github.com/cosmos/ibc-go/blob/2d3181bf37239325002bb6a496a2115b6e88a615/modules/core/04-channel/keeper/handshake.go#L308)

If a relayer changed the version, the proof would fail

Reply:
thanks :pray: